### DISCUSSION TITLE
WPF - Keyboard shortcuts

### DISCUSSION THREAD

**Discussion by bhosachin** [cefsharp/CefSharp#4139] [Q&A]
we have hosted website inside WPF application using CefSharp 85 and it seems keyboard shortcuts of that websites are not working. e.g = to zoom. 
Does anyone know how I can make the Chromium Control behave the same as an external browser in terms of Shortcuts?

**Comment by amaitland** [comment]
It's up to the embedding application to implement keyboard shortcuts. 

There are a number of commands you can directly bind to.

https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.keybinding?view=windowsdesktop-6.0

**Reply by bhosachin** [reply]
Keyboard shortcuts are already implemented in embedded website and those are working when its browsed in external browser but when we host it inside Chromium Control (CefSharp) then those shortcuts keys didn't work at all.
e.g. If we take CefShap minimal example of WPF and host website then shortcuts keys of that sites are not working the ways it works in external browser

**Reply by amaitland** [reply]
Using zoom as an example you need to implement your own key binding to trigger the zoom commands.

**Reply by amaitland** [reply]
Your other option is to provide a custom keyboard handler implementation. 

- http://cefsharp.github.io/api/101.0.x/html/M_CefSharp_Handler_KeyboardHandler_OnKeyEvent.htm
- http://cefsharp.github.io/api/101.0.x/html/P_CefSharp_Wpf_ChromiumWebBrowser_KeyboardHandler.htm
- https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.keyinterop?view=windowsdesktop-6.0

You can directly call SetZoomLevel etc if the key combination matches. 

http://cefsharp.github.io/api/101.0.x/html/T_CefSharp_IBrowser.htm

**Reply by bhosachin** [reply]
Even if we implement keyboard handler, we cant trigger hosted website commands because its abstraction to us and if we change hosted website to something else then again we need implement keyboard handler.
we are expecting, it should work in a way it works in chrome browser.

**Reply by amaitland** [reply]
> we cant trigger hosted website

What do you mean by this? 

> we are expecting, it should work in a way it works in chrome browser

CEF which CefSharp is a wrapper around doesn't provide any shortcuts, it's a no frills implementation. It's up to you to implement. 

You need shortcuts you have to add them.

**Reply by bhosachin** [reply]
Let me try to elaborate problem in details,  we have to host website which is owned and developed by third party and we need to embed it inside our WPF application so we have used cefsharp to host third part website inside our application.
Now if we browse that third party website in chrome browser, then their application shortcut keys works fine but when we host it inside our application using cefsharp then there shortcut keys stops working.

I have also tested it using cefsharp wpf minimal example, where in I have browsed third party URL and also observed shortcut keys are not working

**Reply by amaitland** [reply]
As I said, there are no shortcuts implemented. You will need to implement the ones you wish to support.

In WPF all key events are forwarded to the browser, so you have complete control.

- Implement Key binding if you wish to provide shortcuts at the WPF level.
- Implement `KeyboardHandler` if you want to integrate with the browser more closely.
  - OnPreKeyEvent is called before a keyboard event is sent to the renderer
  - OnKeyEvent is called after the renderer and JavaScript in the page has had a chance to handle the event.

Pick which ever best suites your use case.

**Reply by bhosachin** [reply]
Thanks for the quick reply @amaitland. we don't need to provide shortcuts at WPF level. 
Let me give you one more example, if we host google.com inside our application using Cefsharp and we are expecting that when user press
ctrl+P, it should open print dialog which opens if we browse google.com in Chrome.

This is just an example, now consider instead of google.com we host some other website which is developed and owned by someone else and they have their own custom shortcut keys implemented which works in Chrome but when we host it inside CEF it doesn't work e.g = to zoom to particular area of their sites.

**Reply by amaitland** [reply]
Please confirm you've read both http://cefsharp.github.io/api/101.0.x/html/M_CefSharp_Handler_KeyboardHandler_OnPreKeyEvent.htm and http://cefsharp.github.io/api/101.0.x/html/M_CefSharp_Handler_KeyboardHandler_OnKeyEvent.htm

**Reply by bhosachin** [reply]
Yes, @amaitland, this is how I implemented it. **Pls check my comments in OnKeyEvent**
 
public class KeyboardHandler : IKeyboardHandler
{
    public bool OnPreKeyEvent(IWebBrowser chromiumWebBrowser, IBrowser browser, KeyType type, int windowsKeyCode, int nativeKeyCode, CefEventFlags modifiers, bool isSystemKey, ref bool isKeyboardShortcut)
    {
        isKeyboardShortcut = true;
        return false;
    }

    public bool OnKeyEvent(IWebBrowser chromiumWebBrowser, IBrowser browser, KeyType type, int windowsKeyCode, int nativeKeyCode, CefEventFlags modifiers, bool isSystemKey)
    {
        // what am I supposed to do with this Event now?
        // I do not want to implement shortcut functionality of that website here, since that is part of the web site
        // I just want the website to be able to receive the keystrokes and allow it to handle it
        return false;
    }
}

**Reply by bhosachin** [reply]
@amaitland any leads would be appreciated

**Reply by emily-malone** [reply]
Hey @bhosachin 

I was looped in on this issue on my end (developer on the website developed with hotkeys he is referencing). I started doing a bit of research and stumbled upon this thread. 

My team is reaching out about the version + OS, but in the meantime to keep this investigation moving I just wanted to check - when you do ctrl+P in the cefsharp browser on google.com in the example below does it behave as expected with opening the print page? I am just trying to narrow down possibilities and want to understand if that is an issue as well? 

I have tested on an embedded web browser using lambda test and the keys work as expected there so I am trying to figure out how to reproduce the scenarios. 

> Let me give you one more example, if we host google.com inside our application using Cefsharp and we are expecting that when user press
> ctrl+P, it should open print dialog which opens if we browse google.com in Chrome.

**Comment by emily-malone** [comment]
@amaitland does CefSharp support Proxy?

**Reply by amaitland** [reply]
Yes.

**Reply by bhosachin** [reply]
@amaitland  
We are hosting a site inside WPF Minimal Example which is using vue.js Javascript framework. somehow hotkeys(magic keys) are not working.

We tried to test vue.js magic key usage demo site **https://vueuse.org/core/usemagickeys/** inside WPF minimal example with cefsharp 102

Here is behavior :

WPF minimal example where in vue.js is not able to print hotkeys pressed inside Cefsharp host : 
![image](https://user-images.githubusercontent.com/107455705/176359404-707287af-e42f-4111-bb38-5a30081f9ad5.png)


Browser behavior where its able to print hotkeys are pressed, Highlighted in Red Rectangle :
![image](https://user-images.githubusercontent.com/107455705/176354560-3c9c3fa1-7046-4a69-9299-2def1c383b3a.png)

CC @emily-malone

**Reply by amaitland** [reply]
If I had to guess then I'd say this is likely related to this upstream issue https://bitbucket.org/chromiumembedded/cef/issues/2597/windows-windowless-keyboard-handling

The `DOM Keycode` is not correctly set by `CEF` with windowless (offscreen) rendering which the `WPF` version uses.

An alternative would be to use https://github.com/cefsharp/CefSharp.Wpf.HwndHost

**Reply by bhosachin** [reply]
@amaitland It's working with CefSharp.Wpf.Hwnd but Can we make it working with CefSharp.WPF itself.

**Reply by amaitland** [reply]
You could build your own custom version of CEF with the proposed PR.

**Reply by bhosachin** [reply]
@amaitland 
https://github.com/cefsharp/CefSharp.Wpf.HwndHost has introduced airspace issues in application. Do we have any solution for it.
